Added QEMU/buildx to Github Actions to build multiarch containers

Maka0 opened the pull request:
Had some time so looked into Github Actions some more how buildx could be implemented for #80. It's a bit different then Gitlab CI/drone, but not so bad. Builds take a bit longer though (some python packages for arm64) but should finish in about 15 minutes.

Because it is build/pushed with a manifest now both architectures can use the same tag.

Thanks for creating ytdl-sub!

jmbannon commented on docker/Dockerfile:
```diff
@@ -15,7 +15,11 @@ RUN apk update --no-cache && \
       python3=~3.10 \
-      py3-setuptools && \
+      py3-setuptools \
+      zlib-dev \
```
do these install on the amd64 image?

jmbannon replied:
also lets delete these packages at the end if we can

jmbannon commented on .github/workflows/package.yaml:
```diff
@@ -68,9 +77,10 @@ jobs:
 
+      - name: Build Docker Image and push to registry
+        if: ${{ github.ref }} =~ ^refs\/heads\/master|^refs\/tags\/v.*
+        run: docker buildx build --push --platform=linux/amd64,linux/arm64 --cache-from=type=local,src=/tmp/build-cache --tag ${{ steps.preperations.outputs.IMAGE_TAG }} --label "runnumber=${GITHUB_RUN_ID}" docker/
```
wouldn't this build the image twice when it's on master or a version tag?

jmbannon replied:
ah it's all cached, disregard

Maka0 replied:
Correct it is cached from the previous steps. In theory you could run multiple builds in parallel on multiple machines (don't know how to do this in Github atm due to shared storage etc) and combine them in the final step and push it to the registry. Don't think this will offer much in term of time as buildx is already running these in parallel on a single machine.

jmbannon commented:
Thanks a ton for this! Have a few questions. Overall looks good though

jmbannon commented on .github/workflows/package.yaml:
```diff
@@ -43,8 +43,16 @@ jobs:
           cp dist/*.whl docker/root
 
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
```
Do you think this would affect ffmpeg performance in any way?

Maka0 replied:
It will only affect this during the build stage (hence the slow builds). It sets up a QEMU emulation to emulate arm64 on amd64 (or vice versa if Github would have arm64 runners) so it should be native performance when deployed on the corresponding architecture.

jmbannon commented on .github/workflows/package.yaml:
```diff
@@ -68,9 +77,10 @@ jobs:
+            )"
 
+      - name: Build Docker Image and push to registry
+        if: ${{ github.ref }} =~ ^refs\/heads\/master|^refs\/tags\/v.*
```
Bug here on my part - since your branch name is master, it tried to run this step but failed from bad creds

Maka0 replied:
Any suggestion of how this could be improved? I used this statement of another build step so I don't know when you want to push the container to the registry?

jmbannon replied:
There's probably a way to check the repo owner + name, we should add the check for `jmbannon/ytdl-sub`

jmbannon replied:
I'll try to look into this soon

jmbannon commented:
The build time isn't ideal but I think we can eventually replace Pillow w/ffmpeg. YTDLP uses ffmpeg to convert thumbnails here: https://github.com/yt-dlp/yt-dlp/blob/962ffcf89c8d935410391fbea3580688aafe76d7/yt_dlp/postprocessor/ffmpeg.py#L1077

Maka0 commented:
I've pushed an update to remove the additional packages at the end of the build.

> The build time isn't ideal but I think we can eventually replace Pillow w/ffmpeg. YTDLP uses ffmpeg to convert thumbnails here: https://github.com/yt-dlp/yt-dlp/blob/962ffcf89c8d935410391fbea3580688aafe76d7/yt_dlp/postprocessor/ffmpeg.py#L1077

Can I help with this? Or did you already had a plan for this (I've noticed you're already working on thumbnails)?

jmbannon commented:
> Can I help with this? Or did you already had a plan for this (I've noticed you're already working on thumbnails)?

Help on this would be great. I made an issue for it: https://github.com/jmbannon/ytdl-sub/issues/86
If you don't mind, I'd like to get this done first before merging this PR to keep the docker image build times somewhat fast.

Maka0 commented:
> > Can I help with this? Or did you already had a plan for this (I've noticed you're already working on thumbnails)?
> 
> Help on this would be great. I made an issue for it: #86 If you don't mind, I'd like to get this done first before merging this PR to keep the docker image build times somewhat fast.

Looked into this. Updated my branch to switch Pillow for ffmpeg and optimized the docker build some more (it is now building in parallel). Build time for arm64 went down to 9-10 minutes.
Also changed the example so thumbnail is saved as "-thumb.jpg" so kodi can pick this up.

Maka0 commented:
Should fix #86

jmbannon commented:
>  Also changed the example so thumbnail is saved as "-thumb.jpg" so kodi can pick this up.

Do you have a link by chance to the kodi documentation for this? It works for me without the `-thumb`

jmbannon commented:
Also I got some bad news, I think we need to rebase this branch on latest master. I updated the md5 part of the test fixtures

jmbannon commented on .github/workflows/package.yaml:
```diff
@@ -43,8 +42,129 @@ jobs:
+      packages: write
+      contents: read
+
+    if: contains('refs/heads/master', github.ref) || ${{ github.ref }} =~ ^refs\/tags\/v.*
```
Need to add `(...) && ${GITHUB_REPOSITORY} == 'jmbannon\/ytdl-sub' here so it it skips this at PR time. There might be a more elegant way to do this

jmbannon replied:
It's still running because in your fork, the branch name is master

Maka0 replied:
I don't think we should limit it to the repository then forks wouldn't build anymore. This build is using "refs/remotes/pull/84/merge" so it shouldn't have picked up this final deploy job. Let me check if merge can be ignored.

jmbannon replied:
Good point. I think we can use `(...) && ${{ github.event_name == 'push' }}`

Maka0 replied:
That was my thought as well, but it seemed to be some formatting on the if statement causing it to run too often. Pushed an update to the PR for this.

jmbannon commented:
That did the trick! Thanks again for the PR.
